nano/node/blockprocessor.cpp

```diff
@@ -26,13 +26,60 @@ nano::block_post_events::~block_post_events ()
+	stateless{ node.ledger.constants.epochs },
+	position{},
+	gap_previous{ node.ledger },
+	determinate{ node.ledger },
 	next_log (std::chrono::steady_clock::now ()),
 	node (node_a),
 	write_database_queue (write_database_queue_a),
 	state_block_signature_verification (node.checker, node.ledger.constants.epochs, node.config, node.logger, node.flags.block_processor_verification_size)
 {
-	state_block_signature_verification.blocks_verified_callback = [this] (std::deque<nano::unchecked_info> & items, std::vector<int> const & verifications, std::vector<nano::block_hash> const & hashes, std::vector<nano::signature> const & blocks_signatures) {
-		this->process_verified_state_blocks (items, verifications, hashes, blocks_signatures);
+	// Pipeline begin
+	pipeline = [this] (std::shared_ptr<nano::block> block_a) {
+		stateless.sink (block_a);
+	};
+	stateless.pass = [this] (std::shared_ptr<nano::block> block_a, nano::account const & account) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::stateless_signer_filter_pass);
+		state_block_signature_verification.add ({ block_a, account, nano::signature_verification::unknown });
+	};
+	stateless.reject = [this] (std::shared_ptr<nano::block> block_a) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::stateless_signer_filter_reject);
+		gap_previous.sink (block_a);
+	};
+	gap_previous.pass = [this] (std::shared_ptr<nano::block> block_a, std::shared_ptr<nano::block> previous) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::gap_previous_filter_pass);
+		position.sink (block_a, previous);
+	};
+	gap_previous.reject = [this] (std::shared_ptr<nano::block> block_a) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::gap_previous_filter_reject);
+		node.gap_cache.add (block_a->hash ());
+		node.unchecked.put (block_a->previous (), block_a);
+	};
+	position.pass = [this] (std::shared_ptr<nano::block> block_a, std::shared_ptr<nano::block> previous_a) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::block_position_filter_pass);
+		determinate.sink (block_a, previous_a);
+	};
+	position.reject = [this] (std::shared_ptr<nano::block> block_a) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::block_position_filter_reject);
+	};
+	determinate.pass = [this] (std::shared_ptr<nano::block> block_a, nano::account const & account) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::determinate_signer_filter_pass);
+		state_block_signature_verification.add ({ block_a, account, nano::signature_verification::unknown });
+	};
+	determinate.reject = [this] (std::shared_ptr<nano::block> block_a) {
+		node.stats.inc (nano::stat::type::block_pipeline, nano::stat::detail::determinate_signer_filter_reject);
+		node.workers.push_task ([node = node.shared (), block_a] () {
+			node->unchecked.put (block_a->previous (), block_a);
```

Comment:
Why is this is done by background task when a very similar operation in `gap_previous.reject` is done in the foreground?